crowbar_framework/app/models/api/node.rb

```diff
@@ -50,6 +50,32 @@ def repocheck(options = {})
       end
     end
 
+    def already_upgraded?(node)
+      current_platform = "#{node[:platform]}-#{node[:platform_version]}"
+      return false if current_platform != node[:target_platform]
+
+      Rails.logger.info("node #{node.name} is already upgraded")
+      true
+    end
+
+    def pre_upgrade(node)
+      # Migrate out the l3 agents and shut down pacemaker
+      # FIXME: should we make thise code part of crowbar-upgrade-os.sh?
```

Comment:
I think slightly cleaner to keep it separate, and this might also help later when we make it async and implement an API call for querying the status - then the UI could distinguish between these two steps and show a progress bar or something.